src/lib/core/fiber.h

```diff
@@ -1243,6 +1247,28 @@ fiber_check_gc(void);
 struct lua_State *
 fiber_lua_state(struct fiber *f);
 
+/** Change whether fiber is system or not. */
+static inline void
+fiber_set_system(struct fiber *f, bool yesno)
+{
+	if (yesno) {
+		if (!(f->flags & FIBER_IS_SYSTEM)) {
+			f->flags |= FIBER_IS_SYSTEM;
+			assert(cord()->client_fiber_count > 0);
+			cord()->client_fiber_count--;
```

Comment:
Shouldn't we wake up the shutdown fiber in this case?

Reply:
Fixed.

<details>
<summary>diff</summary>

```diff
diff --git a/src/lib/core/fiber.c b/src/lib/core/fiber.c
index 5c4aed7fe..26bc4459e 100644
--- a/src/lib/core/fiber.c
+++ b/src/lib/core/fiber.c
@@ -2351,6 +2351,26 @@ fiber_lua_state(struct fiber *f)
 	return f->storage.lua.stack;
 }
 
+void
+fiber_set_system(struct fiber *f, bool yesno)
+{
+	if (yesno) {
+		if (!(f->flags & FIBER_IS_SYSTEM)) {
+			f->flags |= FIBER_IS_SYSTEM;
+			assert(cord()->client_fiber_count > 0);
+			cord()->client_fiber_count--;
+			if (cord()->shutdown_fiber != NULL &&
+			    cord()->client_fiber_count == 0)
+				fiber_wakeup(cord()->shutdown_fiber);
+		}
+	} else {
+		if (f->flags & FIBER_IS_SYSTEM) {
+			f->flags &= ~FIBER_IS_SYSTEM;
+			cord()->client_fiber_count++;
+		}
+	}
+}
+
 void
 fiber_shutdown(void)
 {
diff --git a/src/lib/core/fiber.h b/src/lib/core/fiber.h
index f356bc45d..e58553bdd 100644
--- a/src/lib/core/fiber.h
+++ b/src/lib/core/fiber.h
@@ -1248,22 +1248,8 @@ struct lua_State *
 fiber_lua_state(struct fiber *f);
 
 /** Change whether fiber is system or not. */
-static inline void
-fiber_set_system(struct fiber *f, bool yesno)
-{
-	if (yesno) {
-		if (!(f->flags & FIBER_IS_SYSTEM)) {
-			f->flags |= FIBER_IS_SYSTEM;
-			assert(cord()->client_fiber_count > 0);
-			cord()->client_fiber_count--;
-		}
-	} else {
-		if (f->flags & FIBER_IS_SYSTEM) {
-			f->flags &= ~FIBER_IS_SYSTEM;
-			cord()->client_fiber_count++;
-		}
-	}
-}
+void
+fiber_set_system(struct fiber *f, bool yesno);
 
 /** Cancel all client (non system) fibers and wait until they finished. */
 void
diff --git a/test/unit/fiber.cc b/test/unit/fiber.cc
index 82785006e..7caab3a8e 100644
--- a/test/unit/fiber.cc
+++ b/test/unit/fiber.cc
@@ -682,6 +682,17 @@ fiber_test_set_system(void)
 	footer();
 }
 
+static int
+hang_on_cancel_f(va_list ap)
+{
+	while (!fiber_is_cancelled())
+		fiber_yield();
+	fiber_set_system(fiber(), true);
+	while (true)
+		fiber_yield();
+	return 0;
+}
+
 static void
 fiber_test_shutdown(void)
 {
@@ -692,10 +703,13 @@ fiber_test_shutdown(void)
 	fiber_set_joinable(fiber1, true);
 	struct fiber *fiber2 = fiber_new_system("fiber2", wait_cancel_f);
 	fail_unless(fiber2 != NULL);
+	struct fiber *fiber3 = fiber_new("fiber3", hang_on_cancel_f);
+	fail_unless(fiber3 != NULL);
 
 	fiber_shutdown();
 	fail_unless((fiber1->flags & FIBER_IS_DEAD) != 0);
 	fail_unless((fiber2->flags & FIBER_IS_DEAD) == 0);
+	fail_unless((fiber3->flags & FIBER_IS_DEAD) == 0);
 
 	fiber_join(fiber1);
 ```

</details>